### PR TITLE
[RISCV] Combine trunc (sra sext (x), zext (y)) to sra (x, smin (y, scalarsizeinbits(y) - 1))

### PR DESCRIPTION
**Pull request by LWenH**
For RVV, If we want to perform an i8 or i16 element-wise vector arithmetic right shift in the upper C/C++ program,  the value to be shifted would be first sign extended to i32, and the shift amount would also be zero_extended to i32 to perform the vsra.vv instruction, and followed by a truncate to get the final calculation result, such pattern will later expanded to a series of "vsetvli" and "vnsrl" instructions later, this is because the RVV spec only support 2 * SEW -> SEW truncate. But for vector, the shift amount can also be determined by smin (Y, ScalarSizeInBits(Y) - 1)). Also, for the vsra instruction, we only care about the low lg2(SEW) bits as the shift amount.

- Alive2: https://alive2.llvm.org/ce/z/u3-Zdr
- C++ Test cases : https://gcc.godbolt.org/z/q1qE7fbha

### CODE REVIEW

**Comment by LWenH** [comment]
How to add reviewers like in phabricator to request for review in github? I can't find that button in github interfaces(I tried that in the right side "Reviewers" button but that button is not clickable, may be that need some permission in github?). 

@topperc @lukel97  
Request for review, thank you.

**Comment by lukel97** [comment]
> How to add reviewers like in phabricator to request for review in github? I can't find that button in github interfaces(I tried that in the right side "Reviewers" button but that button is not clickable, may be that need some permission in github?).

That's strange, clicking on the "Reviewers" label brings up this menu for me. I don't think you should need to be a member of the llvm-org since it's your own PR?

image

**Comment by LWenH** [comment]
Yeah, I tried that one, but it still unclickable. I even tried to restart the pull request to invoke that lable, but it still doesn't work.  BTW, thank you for helping me to add reviewers.

**Comment by LWenH** [comment]
Ping.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:13686]: Why does N need to have a single use?

Comment (LWenH) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Yeah, I agree with you. There is no need to judge hasOneUse for N here.  Address comment.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by LWenH** [comment]
image
Hi, folks.  I might still not have the write access to the github repository.  Could reviewers merge this pull request for me please? BTW, I want to know how to get the write access for the github repository now？I didn't see the similar description in the new version of the [github submission instructions](https://llvm.org/docs/GitHub.html#github-reviews).  May be also need that write access to get the permission to add reviewers for pull requests in github right now? 

Thank you for your time to read this question.

**Comment by dtcxzyw** [comment]
I will merge this pull request after this patch passes all regression tests on my machine (I cannot rerun the failed workflow on buildkite).
To obtain commit access, please refer to the developer policy:
https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

**Comment by LWenH** [comment]
> I will merge this pull request after this patch passes all regression tests on my machine (I cannot rerun the failed workflow on buildkite). To obtain commit access, please refer to the developer policy: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

Thank you folks, that's a big help for me.

**Pull request merged**
